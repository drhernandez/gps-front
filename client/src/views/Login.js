import React from "react";
import to from "await-to-js";
import { Link, withRouter } from "react-router-dom";
import {
  Container,
  Row,
  Col,
  Card,
  CardBody,
  FormInput,
  FormGroup,
  FormFeedback,
  Form,
  Alert,
} from "shards-react";
import Button from "../components/common/Button";
import { AuthService } from "../api/services";
import validations from "../utils/ValidationsUtil";
import constants from "../utils/Constants";
import store from "../redux/store";
import { setUserInfoAction } from "../redux/actions/actions";
import "../styles/login.css";
var _ = require('lodash');

const errorsDefault = {
  email: {
    required: false
  },
  password: {
    required: false
  }
}

class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      errors: errorsDefault,
      remember: false,
      showAlert: false,
      showSppiner: false
    }

    this.login = this.login.bind(this);
    this.invalidateError.bind(this);
  };

  async login(e) {
    e.preventDefault()
    e.persist();
    const errors = _.clone(errorsDefault);
    errors.email.required = !validations.validateRequired(e.target.email.value);
    errors.password.required = !validations.validateRequired(e.target.password.value);

    if (Object.values(errors).find(fieldValidations => Object.values(fieldValidations).find(value => value))) {  // Si alguno de los errores está en true...
      this.setState({
        errors: errors
      })
    } 
    else {
      this.setState({
        showSppiner: true
      });
      const email = e.target.email.value;
      const pass = e.target.password.value;
      const [err, tokenInfo] = await to(AuthService.login(email, pass));
      if (!err && tokenInfo != null) {
        await store.dispatch(setUserInfoAction(tokenInfo.user));
        this.props.history.push("/");
      } else {
        this.setState({
          showAlert: true,
          showSppiner: false
        })
        setTimeout(() => {
          this.setState({
            showAlert: false
          })
        }, 4000);
      }
    }
  }

  invalidateError(field) {
    let errors = this.state.errors;
    Object.keys(errors[field]).forEach(validationType => errors[field][validationType] = false)
    this.setState({
      errors: errors
    })
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4 h-100">
        <Row noGutters className="h-100">
          <Col className="mx-auto login">
            <Card small>
              <CardBody className="px-4">
                <h5 className="login__titulo text-center mt-5 mb-4">Ingresa a tu cuenta</h5>
                <Form onSubmit={this.login} noValidate>
                  <FormGroup>
                    <label htmlFor="email">Email</label>
                    <FormInput
                      type="email"
                      id="email"
                      placeholder="Dirección de correo"
                      autoComplete="email"
                      invalid={this.state.errors.email.required}
                      onChange={() => this.invalidateError("email")}
                    />
                    <FormFeedback>Campo requerido</FormFeedback>
                  </FormGroup>
                  <FormGroup>
                    <label htmlFor="password">Contraseña</label>
                    <FormInput 
                      id="password" 
                      type="password" 
                      placeholder="Contraseña"
                      invalid={this.state.errors.password.required}
                      onChange={() => this.invalidateError("password")} />
                    <FormFeedback>Campo requerido</FormFeedback>
                  </FormGroup>
                  <FormGroup>
                    <Button type="submit" theme="accent" className="d-table mx-auto" label="Iniciar Sesión" showSppiner={this.state.showSppiner}></Button>
                  </FormGroup>
                  <FormGroup>
                    <Alert className="mb-3" open={this.state.showAlert} theme={constants.Themes.ERROR}>
                      Los datos ingresados no son correctos. Intentá de nuevo
                    </Alert>
                  </FormGroup>
                </Form>
              </CardBody>
            </Card>
            <div className="px-2 pt-3 login__justify_links">
              <Link to="/forgot-password">Olvidaste tu contraseña?</Link>
            </div>
          </Col>
        </Row>
      </Container>
    );
  }
}

export default withRouter(Login);