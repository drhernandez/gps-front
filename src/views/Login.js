import React from "react";
import { Link } from "react-router-dom";
import {
  Container,
  Row,
  Col,
  Card,
  CardBody,
  FormInput,
  FormGroup,
  FormCheckbox,
  Form,
  Button,
} from "shards-react";

import "../styles/login.css";

export default class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      remember: false
    }

    this.toogleCheckbox.bind(this);
    this.login.bind(this);
  };

  toogleCheckbox() {
    this.setState({
      remember: !this.state.remember
    })
  }

  login(e) {
    e.preventDefault()
    e.persist();
    console.log(e.target.email.value);
    console.log(e.target.password.value);
    console.log("remember me: ", this.state.remember)
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4 h-100">
        <Row noGutters className="h-100">
          <Col className="mx-auto login">
            <Card small>
              <CardBody className="px-4">
                {/* <img class="login__logo d-table mx-auto mb-3" src="../images/shards-dashboards-logo.svg" alt="Login Template" /> */}
                <h5 className="login__titulo text-center mt-5 mb-4">Ingresa a tu cuenta</h5>
                <Form onSubmit={e => this.login(e)}>
                  <FormGroup>
                    <label htmlFor="email">Email</label>
                    <FormInput id="email" type="email" placeholder="Email" />
                  </FormGroup>
                  <FormGroup>
                    <label htmlFor="password">Contrase침a</label>
                    <FormInput id="password" type="password" placeholder="Contrase침a" />
                  </FormGroup>
                  <FormGroup>
                    <FormCheckbox checked={this.state.remember} onChange={() => this.toogleCheckbox()}>
                      Recordarme
                    </FormCheckbox>
                  </FormGroup>
                  <Button type="submit" pill className="d-table mx-auto mb-5">Iniciar Sesi칩n</Button>
                </Form>
              </CardBody>
            </Card>
            <div className="px-2 pt-3 login__justify_links">
              <Link to="/forgot-password">Olvidaste tu contrase침a?</Link>
              <Link to="/signup">Crea una cuenta nueva</Link>
            </div>
          </Col>
        </Row>
      </Container>
    );
  }
}