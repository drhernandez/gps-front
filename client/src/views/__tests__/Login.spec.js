import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { MemoryRouter as Router } from "react-router-dom";
import Login from '../Login'
// import { AuthService } from '../../api/services'
// jest.mock("../../api/services/authService")

it("snapshot renders", () => {
  const component = renderer.create(
    <Router>
      <Login />
    </Router>
  );
  expect(component.toJSON()).toMatchSnapshot();
})

it("Link Olvidaste tu contraseÃ±a should redirect to /forgot-password", () => {
  const wrapper = shallow(
    <Login.WrappedComponent />
  );
  
  expect(wrapper.find('Link').at(0).props().to).toBe('/forgot-password')
})

// it("OnSubmit should redirect to home", () => {
//   debugger
//   const historyMock = { push: jest.fn() };
//   const wrapper = mount(
//     <Router>
//       <Login history={historyMock} />
//     </Router>
//   );

//   const event = {
//     email: {
//       value: "test@test.com"
//     },
//     password: {
//       value: "password"
//     }
//   }

//   const mockedEvent = { target: event, preventDefault: () => {}, persist: () => {} }
//   AuthService.login.mockResolvedValue({ user: {} })
//   wrapper.find('Form').prop('onSubmit')(mockedEvent)
//   expect(historyMock.push.mock.calls.length).toBe(1);
// })