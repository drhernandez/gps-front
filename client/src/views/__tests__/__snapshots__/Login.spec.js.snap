// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot renders 1`] = `
<div
  className="main-content-container px-4 h-100 container-fluid"
>
  <div
    className="h-100 no-gutters row"
  >
    <div
      className="mx-auto login col"
    >
      <div
        className="card card-small"
      >
        <div
          className="px-4 card-body"
        >
          <h5
            className="login__titulo text-center mt-5 mb-4"
          >
            Ingresa a tu cuenta
          </h5>
          <form
            className=""
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="email"
              >
                Email
              </label>
              <input
                autoComplete="email"
                className="form-control"
                id="email"
                onChange={[Function]}
                placeholder="Dirección de correo"
                type="email"
              />
              <div
                className="invalid-feedback"
              >
                Campo requerido
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="password"
              >
                Contraseña
              </label>
              <input
                className="form-control"
                id="password"
                onChange={[Function]}
                placeholder="Contraseña"
                type="password"
              />
              <div
                className="invalid-feedback"
              >
                Campo requerido
              </div>
            </div>
            <div
              className="form-group"
            >
              <button
                className="d-table mx-auto btn btn-accent"
                onClick={[Function]}
                type="submit"
              >
                Iniciar Sesión
              </button>
            </div>
            <div
              className="form-group"
            />
          </form>
        </div>
      </div>
      <div
        className="px-2 pt-3 login__justify_links"
      >
        <a
          href="/forgot-password"
          onClick={[Function]}
        >
          Olvidaste tu contraseña?
        </a>
      </div>
    </div>
  </div>
</div>
`;
